dbms/src/Flash/Mpp/MPPTunnelSetWriter.cpp

```diff
@@ -30,52 +30,159 @@ void checkPacketSize(size_t size)
+    {                                                                                       \
+        UPDATE_EXCHANGE_MATRIC_ORIGINAL(type, original_size);                               \
+        switch (compress_method)                                                            \
+        {                                                                                   \
+        case CompressionMethod::NONE:                                                       \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_NONE_COMPRESS(type, is_local, actual_size);              \
+            break;                                                                          \
+        }                                                                                   \
+        case CompressionMethod::LZ4:                                                        \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_LZ4_COMPRESS(type, actual_size);                         \
+            break;                                                                          \
+        }                                                                                   \
+        case CompressionMethod::ZSTD:                                                       \
+        {                                                                                   \
+            UPDATE_EXCHANGE_MATRIC_ZSTD_COMPRESS(type, actual_size);                        \
+            break;                                                                          \
+        }                                                                                   \
+        default:                                                                            \
+            break;                                                                          \
+        }                                                                                   \
+    } while (false)
+
+static inline void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t actual_size, bool is_local)
 {
-    // statistic
-    GET_METRIC(tiflash_exchange_data_bytes, type_hash_original).Increment(packet_bytes);
-    // compression method is always NONE
-    if (is_local)
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(packet_bytes);
-    else
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_remote).Increment(packet_bytes);
+    UPDATE_EXCHANGE_MATRIC(hash, method, original_size, actual_size, is_local);
 }
 
-void updatePartitionWriterMetrics(CompressionMethod method, size_t original_size, size_t sz, bool is_local)
+template <bool is_broadcast, typename FuncIsLocalTunnel, typename FuncWriteToTunnel>
+static void broadcastOrPassThroughWriteImpl(
+    const size_t tunnel_cnt,
```

Comment:
Pass `remote_tunnel_cnt` looks more symmetrical.
```
static void broadcastOrPassThroughWriteImpl(
    const size_t local_tunnel_cnt,
    const size_t remote_tunnel_cnt,
    TrackedMppDataPacketPtr && local_tracked_packet,
    TrackedMppDataPacketPtr && remote_tracked_packet,
    
```

Reply:
It's also an optional way.